Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -30,450 +26,127 @@ field, its finite adèle ring is just defined to be the trivial ring.
+  simp_rw [valuation_of_algebraMap, intValuation_apply, intValuation_lt_one_iff_dvd]
   apply Ideal.finite_factors
   simpa only [Submodule.zero_eq_bot, ne_eq, Ideal.span_singleton_eq_bot]
 
-end IsFiniteAdele
+end IsDedekindDomain
 
-end ProdAdicCompletions
+noncomputable section
 
-open ProdAdicCompletions.IsFiniteAdele
+open Function Set IsDedekindDomain IsDedekindDomain.HeightOneSpectrum
 
-/-- The finite adèle ring of `R` is the restricted product over all maximal ideals `v` of `R`
-of `adicCompletion`, with respect to `adicCompletionIntegers`.
+variable (R : Type*) [CommRing R] [IsDedekindDomain R] {K : Type*}
+    [Field K] [Algebra R K] [IsFractionRing R K]
 
-Note that we make this a `Type` rather than a `Subtype` (e.g., a `Subalgebra`) since we wish
-to endow it with a finer topology than that of the subspace topology. -/
-def FiniteAdeleRing : Type _ := {x : K_hat R K // x.IsFiniteAdele}
+namespace DedekindDomain
 
-namespace FiniteAdeleRing
+variable (R K : Type*) [CommRing R] [IsDedekindDomain R] [Field K] [Algebra R K]
+  [IsFractionRing R K] (v : HeightOneSpectrum R)
+
+open scoped RestrictedProduct
 
-/-- The finite adèle ring of `R`, regarded as a `K`-subalgebra of the
-product of the local completions of `K`.
+/-! ### The finite adèle ring of a Dedekind domain
+We define the finite adèle ring of `R` as the restricted product over all maximal ideals `v` of `R`
+of `adicCompletion` with respect to `adicCompletionIntegers`. We prove that it is a commutative
+ring. -/
 
-Note that this definition exists to streamline the proof that the finite adèles are an algebra
-over `K`, rather than to express their relationship to `K_hat R K` which is essentially a
-detail of their construction.
+/--
+If `K` is the field of fractions of the Dedekind domain `R` then `FiniteAdeleRing R K` is
+the ring of finite adeles of `K`, defined as the restricted product of the completions
+`K_v` with respect to the subrings `R_v`. Here `v` runs through the nonzero primes of `R`
+and the restricted product is the subring of `∏_v K_v` consisting of elements which
+are in `R_v` for all but finitely many `v`.
 -/
-def subalgebra : Subalgebra K (K_hat R K) where
-  carrier := {x : K_hat R K | x.IsFiniteAdele}
-  mul_mem' := mul
-  one_mem' := one
-  add_mem' := add
-  zero_mem' := zero
-  algebraMap_mem' := algebraMap'
+abbrev FiniteAdeleRing : Type _ :=
+  Πʳ v : HeightOneSpectrum R, [v.adicCompletion K, v.adicCompletionIntegers K]
 
-instance : CommRing (FiniteAdeleRing R K) :=
-  Subalgebra.toCommRing (subalgebra R K)
+namespace FiniteAdeleRing
+
+variable {K} in
+/--
+The canonical map from `K` to the finite adeles of `K`. The content of the existence
+of this map is the fact that an element of `K` is integral at all but finitely
+many places, which is `IsDedekindDomain.HeightOneSpectrum.Support.finite R k`.
+For the ring homomorphism, see `DedekindDomain.FiniteAdeleRing.algebraMap`.
+-/
+def algebraMap.toFun (k : K) : FiniteAdeleRing R K :=
```

Comment:
Why not declare this as part of `def AlgebraMap` rather doing it separately on its own?

Reply:
I see you removed the `protected` flag from `AlgebraMap` while implementing this suggestion, was that deliberate? I think it's probably better off with `protected` than without.

Reply:
No it was not deliberate -- well spotted! I wondered why the other proof broke :-)